circleci: failed tests (12955): TestClientDisallowMultipleConns

The following test appears to have failed:

[#12955](https://circleci.com/gh/cockroachdb/cockroach/12955):

```
I0217 22:50:53.755763 235 <autogenerated>:5  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:56047: getsockopt: connection refused"; Reconnecting to "127.0.0.1:56047"
I0217 22:50:59.996801 235 <autogenerated>:5  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:56047: getsockopt: connection refused"; Reconnecting to "127.0.0.1:56047"
I0217 22:51:09.719737 235 <autogenerated>:5  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:56047: getsockopt: connection refused"; Reconnecting to "127.0.0.1:56047"
I0217 22:51:26.288466 235 <autogenerated>:5  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:56047: getsockopt: connection refused"; Reconnecting to "127.0.0.1:56047"
I0217 22:51:50.802674 235 <autogenerated>:5  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:56047: getsockopt: connection refused"; Reconnecting to "127.0.0.1:56047"
panic: test timed out after 1m10s

goroutine 238 [running]:
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:703 +0x132
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x3a

goroutine 1 [chan receive]:
testing.RunTests(0xca6900, 0x10c9d00, 0x1e, 0x1e, 0x412f01)
	/usr/local/go/src/testing/testing.go:562 +0x8ad
testing.(*M).Run(0xc820221ef8, 0x4134d2)
	/usr/local/go/src/testing/testing.go:494 +0x70
github.com/cockroachdb/cockroach/util/leaktest.TestMainWithLeakCheck(0xc820221ef8)
	/go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:37 +0x21
github.com/cockroachdb/cockroach/gossip_test.TestMain(0xc820221ef8)
	/go/src/github.com/cockroachdb/cockroach/gossip/main_test.go:34 +0x21
main.main()
	github.com/cockroachdb/cockroach/gossip/_test/_testmain.go:112 +0x113

goroutine 17 [syscall, 1 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 19 [chan receive]:
github.com/cockroachdb/cockroach/util/log.(*loggingT).flushDaemon(0x10e83c0)
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:1032 +0x64
created by github.com/cockroachdb/cockroach/util/log.init.1
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:613 +0xbc

goroutine 228 [semacquire]:
sync.runtime_Semacquire(0xc82018c094)
	/usr/local/go/src/runtime/sema.go:43 +0x26
sync.(*WaitGroup).Wait(0xc82018c088)
	/usr/local/go/src/sync/waitgroup.go:126 +0xb4
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Stop(0xc82018c070)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:223 +0x1af
github.com/cockroachdb/cockroach/gossip.TestClientDisallowMultipleConns(0xc8200761b0)
	/go/src/github.com/cockroachdb/cockroach/gossip/client_test.go:314 +0x539
testing.tRunner(0xc8200761b0, 0x10c9d60)
	/usr/local/go/src/testing/testing.go:456 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:561 +0x86d

goroutine 235 [sleep]:
time.Sleep(0x92c1f5472)
	/usr/local/go/src/runtime/time.go:59 +0xf9
google.golang.org/grpc.(*Conn).resetTransport(0xc8200da700, 0xc820216d00, 0x0, 0x0)
	/go/src/google.golang.org/grpc/clientconn.go:461 +0x78b
google.golang.org/grpc.NewConn(0xc8200c2b40, 0x4134d2, 0x0, 0x0)
	/go/src/google.golang.org/grpc/clientconn.go:308 +0x441
google.golang.org/grpc.(*unicastPicker).Init(0xc82013faa0, 0xc8200c2b40, 0x0, 0x0)
	/go/src/google.golang.org/grpc/picker.go:75 +0x31
google.golang.org/grpc.Dial(0xc8200ff000, 0xf, 0xc82021fed8, 0x2, 0x2, 0x411817, 0x0, 0x0)
	/go/src/google.golang.org/grpc/clientconn.go:188 +0x27f
--
	/go/src/github.com/cockroachdb/cockroach/gossip/client.go:85 +0x1ba
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc82018c070, 0xc8201894a0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x62
FAIL	github.com/cockroachdb/cockroach/gossip	70.024s
=== RUN   TestParseResolverSpec
--- PASS: TestParseResolverSpec (0.00s)
=== RUN   TestGetAddress
--- PASS: TestGetAddress (0.00s)
PASS
ok  	github.com/cockroachdb/cockroach/gossip/resolver	0.009s
?   	github.com/cockroachdb/cockroach/gossip/simulation	[no test files]
=== RUN   TestKeySorting
--- PASS: TestKeySorting (0.00s)
=== RUN   TestMakeKey

```
Please assign, take a look and update the issue accordingly.

--- a/gossip/client.go
+++ b/gossip/client.go
@@ -21,8 +21,6 @@ import (
 	"net"
 
 	"golang.org/x/net/context"
-	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials"
 
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/rpc"
@@ -70,19 +68,7 @@ func (c *client) start(g *Gossip, disconnected chan *client, ctx *rpc.Context, s
 			disconnected <- c
 		}()
 
-		var dialOpt grpc.DialOption
-		if ctx.Insecure {
-			dialOpt = grpc.WithInsecure()
-		} else {
-			tlsConfig, err := ctx.GetClientTLSConfig()
-			if err != nil {
-				log.Error(err)
-				return
-			}
-			dialOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
-		}
-
-		conn, err := grpc.Dial(c.addr.String(), dialOpt, grpc.WithBlock())
+		conn, err := ctx.GRPCDial(c.addr.String())
 		if err != nil {
 			log.Errorf("failed to dial: %v", err)
 			return

--- a/rpc/rpc.go
+++ b/rpc/rpc.go
@@ -3,6 +3,9 @@ package rpc
 import (
 	"time"
 
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
 	"github.com/cockroachdb/cockroach/base"
 	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/stop"
@@ -37,6 +40,22 @@ func NewContext(context *base.Context, clock *hlc.Clock, stopper *stop.Stopper) 
 		HealthWait:   5 * time.Second,
 	}
 	return ctx
+}
+
+// GRPCDial calls grpc.Dial with the options appropriate for the context.
+func (c *Context) GRPCDial(target string) (*grpc.ClientConn, error) {
+	var dialOpt grpc.DialOption
+	if c.Insecure {
+		dialOpt = grpc.WithInsecure()
+	} else {
+		tlsConfig, err := c.GetClientTLSConfig()
+		if err != nil {
+			return nil, err
+		}
+		dialOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+	}
+
+	return grpc.Dial(target, dialOpt, grpc.WithBlock(), grpc.WithTimeout(base.NetworkTimeout))
 }
 
 // Copy creates a copy of the rpc Context config values, but with a

--- a/server/raft_transport.go
+++ b/server/raft_transport.go
@@ -24,7 +24,6 @@ import (
 	"golang.org/x/net/context"
 
 	"google.golang.org/grpc"
-	"google.golang.org/grpc/credentials"
 
 	"github.com/cockroachdb/cockroach/gossip"
 	"github.com/cockroachdb/cockroach/roachpb"
@@ -136,19 +135,7 @@ func (t *rpcTransport) processQueue(nodeID roachpb.NodeID, storeID roachpb.Store
 		return
 	}
 
-	var dialOpt grpc.DialOption
-	if t.rpcContext.Insecure {
-		dialOpt = grpc.WithInsecure()
-	} else {
-		tlsConfig, err := t.rpcContext.GetClientTLSConfig()
-		if err != nil {
-			log.Error(err)
-			return
-		}
-		dialOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
-	}
-
-	conn, err := grpc.Dial(addr.String(), dialOpt)
+	conn, err := t.rpcContext.GRPCDial(addr.String())
 	if err != nil {
 		log.Errorf("failed to dial: %v", err)
 		return

